Fix <unknown> shape issue for sparse.transpose

yongtang opened the pull request:
This PR tries to address the issue raised in #37638 where
the shape after sparse.transpose is <unknown> even though
the input shape's rank is known. This PR relexed the shape
to only use rank (vs. is_fully_defined) to populate the shape.

The other issue raised about `sparse.reduce_sum` is a little challanging
as the shape is inference from C++ `SparseReduceSumSparse` ops which
will output unknown shape anyway. For that reason this PR does not
address the `sparse.reduce_sum` issue.

This PR fixes #37638.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix <unknown> shape issue for sparse.transpose

This PR tries to address the issue raised in 37638 where
the shape after sparse.transpose is <unknown> even though
the input shape's rank is known. This PR relexed the shape
to only use rank (vs. is_fully_defined) to populate the shape.

The other issue raised about `sparse.reduce_sum` is a little challanging
as the shape is inference from C++ `SparseReduceSumSparse` ops which
will output unknown shape anyway. For that reason this PR does not
address the `sparse.reduce_sum` issue.

This PR fixes 37638.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

mdanatg reviewed (dismissed):
Thank you for the fix! Consider adding a test that verifies the fix.

yongtang committed:
Add test case for <unknown> shape issue for sparse.transpose issue

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@mdanatg Thanks for the review. I have updated the PR to add a test case to cover the change. Since the issue only happens in graph mode (inside tf.function), not eager mode, I have added `@test_util.run_all_in_graph_and_eager_modes` to make sure the test runs in both cases.

Please take a look and let me know if there are any issues.

mdanatg commented on tensorflow/python/framework/sparse_tensor_test.py:
```diff
@@ -59,6 +59,13 @@ def testPythonConstruction(self):
         self.assertAllEqual(sess_run_value.values, value.values)
         self.assertAllEqual(sess_run_value.dense_shape, value.dense_shape)
 
+  @test_util.run_all_in_graph_and_eager_modes
```
I think this decorator was meant to annotate whole classes - I'm not sure it will work. Anyway, if it doesn't, it's always an option to actually create a tf.function and call that instead, something like this:

def testShape(self):
  @def_function.function
  def test_fn():
    tensor = ...
    tensor = ...
    self.assertEqual(tensor.shape.rank, 2)
  test_fn()

yongtang replied:
Ah that works. Thanks!

mdanatg commented on tensorflow/python/framework/sparse_tensor_test.py:
```diff
@@ -59,6 +59,13 @@ def testPythonConstruction(self):
+    tensor = sparse_tensor.SparseTensor(
+        indices=[[0, 0], [1, 2]], values=[1., 2], dense_shape=[3, 4])
+    tensor = sparse_ops.sparse_transpose(tensor)
+    self.assertTrue(tensor.shape.rank == 2)
```
Please use self.assertEqual here.

yongtang replied:
Updated. Thanks!

yongtang committed:
Fix test case based on review

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@mdanatg Thanks for the review! The PR has been updated. Please take a look.
